### ISSUE TITLE
Crash when hold note is moved between columns in mania editor

### ISSUE THREAD

**Issue by ekrctb** [ppy/osu#13297]
Probably known to @smoogipoo but I post here anyway.
Regressed in #12784.

```
System.InvalidOperationException: Cannot call InternalChild unless there's exactly one Drawable in Children (currently 0)!
   at osu.Framework.Graphics.Containers.Container`1.get_Child()
   at osu.Game.Rulesets.Mania.Objects.Drawables.DrawableHoldNote.get_Head() in /data/Projects/Osu/osu/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNote.cs:line 31
   at osu.Game.Rulesets.Mania.Edit.Blueprints.HoldNoteSelectionBlueprint.get_ScreenSpaceSelectionPoint() in /data/Projects/Osu/osu/osu.Game.Rulesets.Mania/Edit/Blueprints/HoldNoteSelectionBlueprint.cs:line 76
   at osu.Game.Screens.Edit.Compose.Components.BlueprintContainer`1.ApplySnapResult(SelectionBlueprint`1[] blueprints, SnapResult result) in /data/Projects/Osu/osu/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs:line 502
   at osu.Game.Screens.Edit.Compose.Components.EditorBlueprintContainer.ApplySnapResult(SelectionBlueprint`1[] blueprints, SnapResult result) in /data/Projects/Osu/osu/osu.Game/Screens/Edit/Compose/Components/EditorBlueprintContainer.cs:line 71
   at osu.Game.Screens.Edit.Compose.Components.BlueprintContainer`1.moveCurrentSelection(DragEvent e) in /data/Projects/Osu/osu/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs:line 498
   at osu.Game.Screens.Edit.Compose.Components.BlueprintContainer`1.OnDrag(DragEvent e) in /data/Projects/Osu/osu/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs:line 189
   at osu.Framework.Graphics.Drawable.TriggerEvent(UIEvent e)
```

**Comment by peppy** [comment]
@smoogipoo you going to look into this one or shall i?

**Comment by smoogipoo** [comment]
I've briefly looked into it but I'm not immediately sure how to fix the issue, so I'm not going to continue for now.

tl;dr: Move selection -> Defaults applied -> OnApply() -> free() -> clears nested hitobjects.

Regressed in https://github.com/ppy/osu/pull/12916 as a result of calling Beatmap.Update() on start time changes.

**Comment by ekrctb** [comment]
When a selected note is moved between columns, the hit object is removed from the playfield:

https://github.com/ppy/osu/blob/b5afe4e50608d8f8a2144ce599e71246f2f1145a/osu.Game.Rulesets.Mania/Edit/ManiaSelectionHandler.cs#L61-L63

And DHO will be removed:

<https://github.com/ppy/osu/blob/ff0494229524c7c5d984895f8500c31b8329185e/osu.Game/Rulesets/UI/HitObjectContainer.cs#L152-L156>

The removal of DHO from the `HitObejctContainer` immediately makes the DHO to be returned to the pool. The DHO nested hit objects are cleared:

https://github.com/ppy/osu/blob/40949f6c1b0c1088f4c28b94b8f9c30084952f69/osu.Game/Rulesets/Objects/Pooling/PoolableDrawableWithLifetime.cs#L91-L98

The `HitObjectUsageFinished` event above is supposed to remove corresponding blueprints. However, the blueprint removal is queued to the `HitObjectUsageEventBuffer`.
Thus, during the frame, the selection blueprint has a stale reference to the DHO.
The crash is due to the usage of this stale DHO.

Because `HitObjectUsageEventBuffer` is just introduced to this delayed blueprint removal, the remaining way to eliminate the stale DHO is to delay the returning of the DHO to the pool, somehow.

**Comment by bdach** [comment]
I tried to see whether it's possible to compute the selection point without using the DHO reference, and it is and it fixes the crash, but the computation is pretty gnarly...

```csharp
        public override Vector2 ScreenSpaceSelectionPoint
        {
            get
            {
                var column = composer.Playfield.GetColumn(HitObject.Column);
                var hoc = column.HitObjectArea.HitObjectContainer;
                var scrollingPosition = scrollingInfo.Algorithm.PositionAt(HitObject.StartTime, column.Time.Current, scrollingInfo.TimeRange.Value, hoc.DrawHeight);
                var correctedPosition = scrollingInfo.Direction.Value == ScrollingDirection.Up ? scrollingPosition : hoc.DrawHeight - scrollingPosition;
                return hoc.ToScreenSpace(new Vector2(hoc.DrawWidth / 2, correctedPosition));
            }
        }
```

Not sure if too much to be tolerable... It seems like too much. Maybe this can be encapsulated better in `Column` or something, not sure.

**Comment by smoogipoo** [comment]
In the context of mania I don't think that's bad, but is that really what we want to do? From the context of arbitrary rulesets I mean...

**Comment by bdach** [comment]
That's a fair call. I only thought of exploring this direction upon seeing this comment:

https://github.com/ppy/osu/blob/3cd7d45b0fa3c15d7ceac2deb2cd94e469b41bb6/osu.Game.Rulesets.Mania/Edit/Blueprints/HoldNoteSelectionBlueprint.cs#L61

which made me think that referencing DHOs from blueprints was unwanted. May have misunderstood intent (that the `.IsLoaded` part is the problem).

**Comment by smoogipoo** [comment]
I still think that is a good direction - the osu! blueprints were changed to refer to the `HitObject` in their `Update()` but unfortunately still use the DHO for these properties too... I would eventually like to see `DrawableObject` removed from the blueprints.

My concern here is around the overhead for something that's required for every ruleset with multiple playfields. Maybe it's the rest of the code related to moving hitobjects around that needs greater unification instead.

**Comment by bdach** [comment]
Yeah I don't know about doing that just for tests, may be worth just changing the tests. Long shot as I don't have all the context available but maybe `IPositionSnapProvider` can be leveraged somehow? Even if by implementing a mock or something.
